### PR TITLE
Highlighting: more C++ improvements

### PR DESCRIPTION
**Pull request by olafurpg**
* Treat raw string literals as strings
* Treat nullptr as a null literal

I was hoping to be able to use `@string.escape` for the character sequence before in raw string literals but the grammar we're using (official C++ grammar) doesn't provide AST nodes for them. We should consider moving to a non-official grammar that supports these, see https://sourcegraph.slack.com/archives/C03C5EFRMR8/p1678881469547489

## Test plan

See updated snapshots.

### CODE REVIEW

File: docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/files/cpp_example3.cc [hunk 1]
```diff
@@ -107,9 +107,23 @@ void literals() {
...
-    World)";
+    auto utf8StringLit = u8"World";
+    auto utf16StringLit = u"World";
+    auto utf32StringLit = U"World";
+    const wchar_t* sC = LR"--(STUV)--"; // ok, raw string literal
+    auto rawStringLit = R"sequence(Hello
+    \n
+    \r
+    \t
+    World)sequence";
```
Comment (varungandhi-src) [inline docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/files/cpp_example3.cc:118]: Maybe add a link to https://en.cppreference.com/w/cpp/language/string_literal here? (I'm assuming this code was taken from there.)

File: docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/sg_syntax__sg_treesitter__test__cpp_example3.cc.snap [hunk 2]
```diff
@@ -362,13 +362,77 @@ expression: "dump_document(&document, &contents)"
...
+//                          ^^^^^^^^ StringLiteral 
+      auto utf32StringLit = U"World";
+//    ^^^^ Keyword 
+//         ^^^^^^^^^^^^^^ Identifier 
+//                          ^^^^^^^^ StringLiteral 
+      const wchar_t* sC = LR"--(STUV)--"; // ok, raw string literal
+//    ^^^^^ Keyword 
+//          ^^^^^^^ IdentifierType 
+//                   ^^ Identifier 
+//                        ^^^^^^^^^^^^^^ StringLiteral
```
Comment (varungandhi-src) [inline docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/sg_syntax__sg_treesitter__test__cpp_example3.cc.snap]: FWIW, I like CLion's coloring scheme here compared to VS Code's (which just uses the same color for everything).

image


It makes it clear that the content of the string is what's inside the brackets, whereas the rest is just syntax.

If possible, would be cool to make this more fine-grained.

**Review by varungandhi-src — APPROVED** [review]

File: docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/sg_syntax__sg_treesitter__test__cpp_example3.cc.snap (hunk 2 above)
Comment (varungandhi-src) [reply docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/sg_syntax__sg_treesitter__test__cpp_example3.cc.snap]: Ah got it. I didn't quite understand the point in the PR description earlier, but that makes sense.

Comment (olafurpg) [reply docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/sg_syntax__sg_treesitter__test__cpp_example3.cc.snap]: The sentence was half-complete 😅

**Pull request merged**